File: src/hlhandler/client/vault.py
"""Vault API client for Hyperliquid."""

from decimal import Decimal
from typing import Any

from hlhandler.client.base import APIError, BaseClient
from hlhandler.config import NetworkConfig
from hlhandler.models import TradingSignal, VaultDetails, VaultInfo, VaultPosition
from hlhandler.signer import Signer


class VaultNotFoundError(APIError):
    """Vault not found."""

    pass


class LockupPeriodError(APIError):
    """Cannot withdraw during lockup period."""

    pass


class VaultClient(BaseClient):
    """Client for Hyperliquid Vault API."""

    def __init__(
        self,
        network: NetworkConfig,
        signer: Signer | None = None,
        **kwargs,
    ):
        """Initialize the vault client.

        Args:
            network: Network configuration.
            signer: Optional signer for authenticated operations.
        """
        super().__init__(network, **kwargs)
        self.signer = signer

    async def list_vaults(
        self,
        min_tvl: Decimal | None = None,
        min_apr: Decimal | None = None,
    ) -> list[VaultInfo]:
        """List public vaults.

        Args:
            min_tvl: Minimum TVL filter.
            min_apr: Minimum APR filter.

        Returns:
            List of VaultInfo.
        """
        result = await self._post("info", {"type": "vaults"})

        vaults = []
        for vault_data in result:
            vault = self._parse_vault_info(vault_data)

            # Apply filters
            if min_tvl is not None and vault.tvl < min_tvl:
                continue
            if min_apr is not None and vault.apr < min_apr:
                continue

            vaults.append(vault)

        return vaults

    async def get_vault_details(
        self,
        vault_address: str,
        user_address: str | None = None,
    ) -> VaultDetails:
        """Get detailed information about a vault.

        Args:
            vault_address: The vault address.
            user_address: Optional user address to get follower state.

        Returns:
            VaultDetails with portfolio and optionally follower state.

        Raises:
            VaultNotFoundError: If vault doesn't exist.
        """
        request: dict[str, Any] = {"type": "vaultDetails", "vault": vault_address}
        if user_address:
            request["user"] = user_address

        try:
            result = await self._post("info", request)
        except APIError as e:
            if "not found" in str(e).lower():
                raise VaultNotFoundError(f"Vault not found: {vault_address}")
            raise

        return self._parse_vault_details(result)

    async def deposit_to_vault(
        self,
        vault_address: str,
        amount_usd: Decimal,
    ) -> bool:
        """Deposit USD to a vault.

        Args:
            vault_address: The vault address.
            amount_usd: Amount to deposit in USD.

        Returns:
            True if successful.

        Raises:
            ValueError: If signer not configured.
        """
        if self.signer is None:
            raise ValueError("Signer required for vault operations")

        action = {
            "type": "vaultDeposit",
            "vault": vault_address,
            "usd": str(amount_usd),
        }

        payload = self.signer.sign_action(action)

        try:
            result = await self._post("exchange", payload)
            return result.get("status") == "ok"
        except Exception:
            return False

    async def withdraw_from_vault(
        self,
        vault_address: str,
        shares: Decimal,
    ) -> bool:
        """Withdraw from a vault by shares.

        Args:
            vault_address: The vault address.
            shares: Amount of shares to withdraw (0-1 for percentage).

        Returns:
            True if successful.

        Raises:
            ValueError: If signer not configured.
            LockupPeriodError: If in lockup period.
        """
        if self.signer is None:
            raise ValueError("Signer required for vault operations")

        action = {
            "type": "vaultWithdraw",
            "vault": vault_address,
            "shares": str(shares),
        }

        payload = self.signer.sign_action(action)

        try:
            result = await self._post("exchange", payload)
            return result.get("status") == "ok"
        except APIError as e:
            if "lockup" in str(e).lower():
                raise LockupPeriodError(str(e))
            return False
        except Exception:
            return False

    async def get_my_vault_positions(
        self,
        user_address: str | None = None,
    ) -> list[VaultPosition]:
        """Get user's positions in all vaults.

        Args:
            user_address: User address. Uses signer address if not provided.

        Returns:
            List of VaultPosition.
        """
        if user_address is None:
            if self.signer is None:
                raise ValueError("User address or signer required")
            user_address = self.signer.address

        result = await self._post(
            "info",
            {"type": "userVaultEquities", "user": user_address},
        )

        positions = []
        for pos_data in result:
            positions.append(
                VaultPosition(
                    vault=pos_data.get("vault", ""),
                    vault_name=pos_data.get("vaultName", ""),
                    shares=Decimal(str(pos_data.get("shares", "0"))),
                    deposited=Decimal(str(pos_data.get("deposited", "0"))),
                    current_value=Decimal(str(pos_data.get("currentValue", "0"))),
                )
            )

        return positions

    async def create_vault(
        self,
        name: str,
        description: str = "",
        is_public: bool = True,
        max_capacity: Decimal | None = None,
        lockup_period: int = 86400,  # 24 hours
        profit_share: Decimal = Decimal("10"),
    ) -> str:
        """Create a new vault.

        Args:
            name: Vault name.
            description: Vault description.
            is_public: Whether the vault is public.
            max_capacity: Maximum capacity in USD.
            lockup_period: Lockup period in seconds.
            profit_share: Leader's profit share percentage.

        Returns:
            New vault address.

        Raises:
            ValueError: If signer not configured.
        """
        if self.signer is None:
            raise ValueError("Signer required for vault operations")

        action: dict[str, Any] = {
            "type": "createVault",
            "name": name,
            "description": description,
            "isPublic": is_public,
            "lockupPeriod": lockup_period,
            "profitShare": str(profit_share),
        }

        if max_capacity is not None:
            action["maxCapacity"] = str(max_capacity)

        payload = self.signer.sign_action(action)
        result = await self._post("exchange", payload)

        if result.get("status") == "ok":
            return result.get("response", {}).get("vault", "")

        raise APIError(result.get("response", "Failed to create vault"))

    def _parse_vault_info(self, data: dict) -> VaultInfo:
        """Parse vault info from API response."""
        return VaultInfo(
            address=data.get("vault", ""),
            name=data.get("name", ""),
            leader=data.get("leader", ""),
            tvl=Decimal(str(data.get("tvl", "0"))),
            apr=Decimal(str(data.get("apr", "0"))),
            profit_share=Decimal(str(data.get("profitShare", "0"))),
            lockup_period=int(data.get("lockupPeriod", 0)),
            is_public=data.get("isPublic", True),
            followers=int(data.get("followers", 0)),
            max_capacity=self.to_decimal(data.get("maxCapacity")),
        )

    def _parse_vault_details(self, data: dict) -> VaultDetails:
        """Parse vault details from API response."""
        info = self._parse_vault_info(data)

        portfolio = data.get("portfolio", {})
        account_value = Decimal(str(portfolio.get("accountValue", "0")))
        positions = portfolio.get("positions", [])

        follower_state = data.get("followerState")

        return VaultDetails(
            info=info,
            account_value=account_value,
            positions=positions,
            follower_state=follower_state,
        )

File: internal/client/vault.go
package client

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/shopspring/decimal"

	"github.com/skilus/hyperhandler/internal/config"
	"github.com/skilus/hyperhandler/internal/models"
	"github.com/skilus/hyperhandler/internal/signer"
)

// errSignerRequired is returned by vault write operations when no signer is set.
var errSignerRequired = errors.New("signer required for vault operations")

// VaultClient drives the Hyperliquid Vault API. The signer is optional: it is
// required only for authenticated operations (deposit/withdraw/create). Mirrors
// client/vault.py:VaultClient.
type VaultClient struct {
	*BaseClient
	signer *signer.Signer
}

// NewVaultClient builds a VaultClient. signer may be nil for read-only use.
func NewVaultClient(network config.NetworkConfig, s *signer.Signer, opts ...Option) *VaultClient {
	return &VaultClient{BaseClient: NewBaseClient(network, opts...), signer: s}
}

// vaultWire is the raw vault summary/details object.
type vaultWire struct {
	Vault        string           `json:"vault"`
	Name         string           `json:"name"`
	Leader       string           `json:"leader"`
	TVL          decimal.Decimal  `json:"tvl"`
	APR          decimal.Decimal  `json:"apr"`
	ProfitShare  decimal.Decimal  `json:"profitShare"`
	LockupPeriod int              `json:"lockupPeriod"`
	IsPublic     *bool            `json:"isPublic"`
	Followers    int              `json:"followers"`
	MaxCapacity  *decimal.Decimal `json:"maxCapacity"`
}

func (w vaultWire) toInfo() models.VaultInfo {
	isPublic := true
	if w.IsPublic != nil {
		isPublic = *w.IsPublic
	}
	return models.VaultInfo{
		Address:      w.Vault,
		Name:         w.Name,
		Leader:       w.Leader,
		TVL:          w.TVL,
		APR:          w.APR,
		ProfitShare:  w.ProfitShare,
		LockupPeriod: w.LockupPeriod,
		IsPublic:     isPublic,
		Followers:    w.Followers,
		MaxCapacity:  w.MaxCapacity,
	}
}

// ListVaults lists public vaults, optionally filtered by minimum TVL/APR (nil to
// skip a filter). Mirrors vault.py:list_vaults.
func (c *VaultClient) ListVaults(ctx context.Context, minTVL, minAPR *decimal.Decimal) ([]models.VaultInfo, error) {
	raw, err := c.post(ctx, "info", map[string]any{"type": "vaultSummaries"}, true)
	if err != nil {
		return nil, err
	}
	var wire []vaultWire
	if err := json.Unmarshal(raw, &wire); err != nil {
		return nil, newAPIError(fmt.Sprintf("decode vault summaries: %v", err), 0, nil)
	}
	vaults := make([]models.VaultInfo, 0, len(wire))
	for _, w := range wire {
		v := w.toInfo()
		if minTVL != nil && v.TVL.Cmp(*minTVL) < 0 {
			continue
		}
		if minAPR != nil && v.APR.Cmp(*minAPR) < 0 {
			continue
		}
		vaults = append(vaults, v)
	}
	return vaults, nil
}

// GetVaultDetails fetches a vault's details, optionally with the follower state
// for userAddress (empty to skip). Returns *VaultNotFoundError if the vault is
// unknown. Mirrors vault.py:get_vault_details.
func (c *VaultClient) GetVaultDetails(ctx context.Context, vaultAddress, userAddress string) (*models.VaultDetails, error) {
	req := map[string]any{"type": "vaultDetails", "vault": vaultAddress}
	if userAddress != "" {
		req["user"] = userAddress
	}

	raw, err := c.post(ctx, "info", req, true)
	if err != nil {
		if e, ok := asAPIError(err); ok && strings.Contains(strings.ToLower(e.Message), "not found") {
			return nil, &VaultNotFoundError{newAPIError("Vault not found: "+vaultAddress, 0, nil)}
		}
		return nil, err
	}

	var data struct {
		vaultWire
		Portfolio struct {
			AccountValue decimal.Decimal  `json:"accountValue"`
			Positions    []map[string]any `json:"positions"`
		} `json:"portfolio"`
		FollowerState map[string]any `json:"followerState"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, newAPIError(fmt.Sprintf("decode vault details: %v", err), 0, nil)
	}

	return &models.VaultDetails{
		Info:          data.toInfo(),
		AccountValue:  data.Portfolio.AccountValue,
		Positions:     data.Portfolio.Positions,
		FollowerState: data.FollowerState,
	}, nil
}

// DepositToVault deposits USD into a vault. Requires a signer. Mirrors
// vault.py:deposit_to_vault.
func (c *VaultClient) DepositToVault(ctx context.Context, vaultAddress string, amountUSD decimal.Decimal) (bool, error) {
	if c.signer == nil {
		return false, errSignerRequired
	}
	action := signer.NewOrderedMap(
		"type", "vaultDeposit",
		"vault", vaultAddress,
		"usd", amountUSD.String(),
	)
	payload, err := c.signer.SignAction(action, nonceNow(), nil, nil)
	if err != nil {
		return false, nil
	}
	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		return false, nil
	}
	return statusOK(raw), nil
}

// WithdrawFromVault withdraws shares from a vault. Requires a signer. Returns
// *LockupPeriodError if the vault is still in lockup. Mirrors vault.py:withdraw_from_vault.
func (c *VaultClient) WithdrawFromVault(ctx context.Context, vaultAddress string, shares decimal.Decimal) (bool, error) {
	if c.signer == nil {
		return false, errSignerRequired
	}
	action := signer.NewOrderedMap(
		"type", "vaultWithdraw",
		"vault", vaultAddress,
		"shares", shares.String(),
	)
	payload, err := c.signer.SignAction(action, nonceNow(), nil, nil)
	if err != nil {
		return false, nil
	}
	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		if e, ok := asAPIError(err); ok && strings.Contains(strings.ToLower(e.Message), "lockup") {
			return false, &LockupPeriodError{newAPIError(e.Message, 0, nil)}
		}
		return false, nil
	}
	return statusOK(raw), nil
}

// GetMyVaultPositions returns the user's positions across all vaults. If
// userAddress is empty it uses the signer's address (signer required then).
// Mirrors vault.py:get_my_vault_positions.
func (c *VaultClient) GetMyVaultPositions(ctx context.Context, userAddress string) ([]models.VaultPosition, error) {
	if userAddress == "" {
		if c.signer == nil {
			return nil, errors.New("user address or signer required")
		}
		userAddress = c.signer.Address()
	}

	raw, err := c.post(ctx, "info", map[string]any{"type": "userVaultEquities", "user": userAddress}, true)
	if err != nil {
		return nil, err
	}
	var wire []struct {
		Vault        string          `json:"vault"`
		VaultName    string          `json:"vaultName"`
		Shares       decimal.Decimal `json:"shares"`
		Deposited    decimal.Decimal `json:"deposited"`
		CurrentValue decimal.Decimal `json:"currentValue"`
	}
	if err := json.Unmarshal(raw, &wire); err != nil {
		return nil, newAPIError(fmt.Sprintf("decode vault equities: %v", err), 0, nil)
	}
	positions := make([]models.VaultPosition, 0, len(wire))
	for _, w := range wire {
		positions = append(positions, models.VaultPosition{
			Vault:        w.Vault,
			VaultName:    w.VaultName,
			Shares:       w.Shares,
			Deposited:    w.Deposited,
			CurrentValue: w.CurrentValue,
		})
	}
	return positions, nil
}

// CreateVault creates a new vault and returns its address. Requires a signer.
// Mirrors vault.py:create_vault.
func (c *VaultClient) CreateVault(
	ctx context.Context,
	name, description string,
	isPublic bool,
	maxCapacity *decimal.Decimal,
	lockupPeriod int,
	profitShare decimal.Decimal,
) (string, error) {
	if c.signer == nil {
		return "", errSignerRequired
	}
	action := signer.NewOrderedMap(
		"type", "createVault",
		"name", name,
		"description", description,
		"isPublic", isPublic,
		"lockupPeriod", lockupPeriod,
		"profitShare", profitShare.String(),
	)
	if maxCapacity != nil {
		action.Set("maxCapacity", maxCapacity.String())
	}

	payload, err := c.signer.SignAction(action, nonceNow(), nil, nil)
	if err != nil {
		return "", err
	}
	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		return "", err
	}
	var resp struct {
		Status   string `json:"status"`
		Response struct {
			Vault string `json:"vault"`
		} `json:"response"`
	}
	if json.Unmarshal(raw, &resp) != nil || resp.Status != "ok" {
		return "", newAPIError("Failed to create vault", 0, nil)
	}
	return resp.Response.Vault, nil
}
